File: spark/etl-streaming-preprocess/src/main/scala/me/w1992wishes/spark/streaming/preprocess/util/KafkaSender.java
package me.w1992wishes.spark.streaming.preprocess.util;

import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerRecord;

import java.io.Serializable;
import java.util.Properties;

public class KafkaSender implements Serializable {

    private static KafkaSender instance = null;

    private KafkaProducer<String, String> producer;

    private KafkaSender(Properties properties) {
        this.producer = new KafkaProducer<String, String>(properties);
    }

    public static synchronized KafkaSender getInstance(Properties properties) {
        if (instance == null) {
            instance = new KafkaSender(properties);
        }
        return instance;
    }

    // 单条发送
    public void send(String topic, String message) {
        producer.send(new ProducerRecord<String, String>(topic, message));
    }

    public void shutdown() {
        producer.close();
    }


}

File: my-common/src/main/scala/me/w1992wishes/my/common/util/KafkaSender.scala
package me.w1992wishes.my.common.util

import java.util.concurrent.Future

import org.apache.kafka.clients.producer.{KafkaProducer, ProducerRecord, RecordMetadata}

/**
  * 使用 懒加载 创建 Kafka Sender
 *
  * @param createProducer 构造 KafkaProducer
  * @tparam K key
  * @tparam V value
  */
class KafkaSender[K, V](createProducer: () => KafkaProducer[K, V]) extends Serializable {
  lazy val producer = createProducer()

  def send(topic: String, key: K, value: V): Future[RecordMetadata] = {
    producer.send(new ProducerRecord[K, V](topic, key, value))
  }

  def send(topic: String, value: V): Future[RecordMetadata] = {
    producer.send(new ProducerRecord[K, V](topic, value))
  }
}

object KafkaSender {

  import scala.collection.JavaConversions._

  def apply[K, V](config: Map[String, Object]): KafkaSender[K, V] = {
    val createProducerFunc = () => {
      val producer = new KafkaProducer[K, V](config)
      sys.addShutdownHook {
        producer.close()
      }
      producer
    }
    new KafkaSender(createProducerFunc)
  }

  def apply[K, V](config: java.util.Properties): KafkaSender[K, V] = apply(config.toMap)
}